src/main/proto/netflix/titus/titus_cluster_membership_api.proto

```diff
@@ -0,0 +1,205 @@
+    // set to `NonLeader`.
+    LeadershipState leadershipState = 3;
+
+    repeated ClusterMemberAddress addresses = 4;
+
+    // Collection of key/value pairs. Read-only special name spaces:
+    // * 'aws.instanceGroupId' - member's ASG instance group id
+    // * 'aws.deployment.XXX' - deployment data (ASG creation time, EC2 instance launch time, etc)
+    // * 'aws.asgTags.XXX' - service's ASG tags
+    // * 'aws.ec2Tags.XXX' - service's EC2 tags
+    // * 'release.XXX' - release data (number, build time, bake time)
+    map<string, string> labels = 5;
+}
+
+message ClusterMembershipRevision {
+
+    ClusterMember current = 1;
+
+    // (Required) Detailed code for being in a specific state.
+    string code = 2;
+
+    // (Optional) Textual description accompanying the 'reasonCode'.
+    string message = 3;
+
+    // Time when a transition to a state happened.
+    uint64 timestamp = 4;
+}
+
+message ClusterMembershipEvent {
+
+    message Snapshot {
+        repeated ClusterMembershipRevision revisions = 1;
+    }
+
+    message Updated {
+        ClusterMembershipRevision revision = 1;
+    }
+
+    message Removed {
+        ClusterMembershipRevision revision = 1;
+    }
+
+    oneof Event {
+        Snapshot snapshot = 1;
+        Updated updated = 2;
+        Removed removed = 3;
+    }
+}
+
+// ----------------------------------------------------------------------------
+// GRPC services
+
+message MemberId {
+    string id = 1;
+}
+
+message ClusterMembers {
+    repeated ClusterMember members = 1;
+}
+
+message QueryRequest {
+    // Filtering criteria:
+    // * 'ids' - comma separated list of instance ids
+    // * 'instanceLimit' - maximum number of records to return per instance
+    // * 'limit' - maximum number of records to return
+    // * 'before' - events before the given timestamp (epoch)
+    // * 'since' - events at or after the given timestamp (epoch)
+    map<string, string> filteringCriteria = 1;
+
+    /// (Optional) If set, only field values explicitly specified in this parameter will be returned.
+    repeated string fields = 2;
+}
+
+message ClusterMembershipRevisions {
+    repeated ClusterMembershipRevision revisions = 1;
+}
+
+message UpdateMemberLabelsRequest {
+    string memberId = 1;
+
+    map<string, string> labels = 2;
+}
+
+message DeleteMemberLabelsRequest {
+    string memberId = 1;
+    repeated string keys = 2;
+}
+
+message UpdateMemberStateRequest {
+    string memberId = 1;
+
+    string reasonMessage = 2;
+
+    ClusterMember.State newState = 3;
+}
+
+service ClusterMembershipService {
+
+    // Get all known cluster members.
+    rpc GetMembers (google.protobuf.Empty) returns (ClusterMembers) {
+    }
+
+    // Get member with the given id.
+    rpc GetMember (MemberId) returns (ClusterMember) {
+    }
+
+    // Get recent changes.
+    rpc GetMembershipRevisions (QueryRequest) returns (ClusterMembershipRevisions) {
```

Comment:
will this return (at least) one revision per memberId (e.g.: the last revision for each member), or only the ones recently changed? If the latter, how is "being recent" defined?

Reply:
This is defined by `QueryRequest`. It is not limited to recent changes.